crates/ethereum-rpc/src/lib.rs

```diff
@@ -638,52 +565,135 @@ fn register_rpc_methods<C: sov_modules_api::Context, Da: DaService>(
 
             let opts: Option<GethDebugTracingOptions> = params.optional_next()?;
 
-            // If opts is None or if opts.tracer is None, then do not check cache or insert cache, just perform the operation
-            // also since this is not cached we need to stop at somewhere, so we add param stop_at
-            if opts.as_ref().map_or(true, |o| o.tracer.is_none()) {
-                let traces = evm.trace_block_transactions_by_number(
-                    block_number,
-                    opts.clone(),
-                    Some(trace_index as usize),
-                    &mut working_set,
-                )?;
-                return Ok(traces[trace_index as usize].clone());
-            }
-
-            // check cache if found convert to requested tracer and config and return
-            if let Some(traces) = ethereum.trace_cache.lock().unwrap().get(&block_number) {
-                let requested_opts = opts.clone().unwrap();
-                let tracer_config = requested_opts.tracer_config;
-                let traces = get_traces_with_reuqested_tracer_and_config(
-                    vec![traces[trace_index as usize].clone()],
-                    requested_opts.tracer.unwrap(),
-                    tracer_config,
-                )?;
-                return Ok(traces.into_iter().next().unwrap());
-            }
-
-            let cache_options = create_trace_cache_opts();
-            let traces = evm.trace_block_transactions_by_number(
+            let traces = debug_trace_by_block_number(
                 block_number,
-                Some(cache_options),
-                None,
+                Some(trace_idx as usize),
+                &ethereum,
+                &evm,
                 &mut working_set,
+                opts,
             )?;
-            ethereum
-                .trace_cache
-                .lock()
-                .unwrap()
-                .insert(block_number, traces.clone());
-            // Convert the traces to the requested tracer and config
-            let requested_opts = opts.clone().unwrap();
-            let tracer_config = requested_opts.tracer_config;
-            let traces = get_traces_with_reuqested_tracer_and_config(
-                vec![traces[trace_index as usize].clone()],
-                requested_opts.tracer.unwrap(),
-                tracer_config,
-            )?;
+            Ok(traces[0].clone())
+        },
+    )?;
 
-            Ok(traces.into_iter().next().unwrap())
+    rpc.register_subscription(
+        "debug_subscribe",
+        "debug_subscription",
+        "debug_unsubscribe",
```

Comment:
How is `unsubscribe` handled here?

Reply:
Once the client sends a rpc request with method `debug_unsubscribe` and the subscription id as the `params`, jsonrpsee will automatically unsubscribe the client.

For this endpoint, even though this is a subscription endpoint due to potentially really long operation, it is finite because start and end blocks are specified, hence it makes sense for this subscription or connection to be closed by the server once the last block is sent. But it appears that jsonrpsee does not provide closing functionality on the server side.

It makes sense for server to not close for real-time endpoints, but for this case I believe it should be done. But such functionality simply doesn't exist for some reason.